packages/wrangler/src/__tests__/deploy.test.ts

```diff
@@ -4422,6 +4442,243 @@ addEventListener('fetch', event => {});`
+					async ({ request }) => {
+						bodies.push(await request.json());
+						return HttpResponse.json(
+							{
+								success: true,
+								errors: [],
+								messages: [],
+								result: { jwt: "<<aus-completion-token>>", buckets: [] },
+							},
+							{ status: 201 }
+						);
+					}
+				)
+			);
+			// skips asset uploading since empty buckets returned
+			mockSubDomainRequest();
+			mockUploadWorkerRequest({
+				expectedExperimentalAssetsForm: [
+					["metadata", JSON.stringify({ assets: "<<aus-completion-token>>" })],
+				],
+			});
+			await runWrangler(
+				"deploy --name test-name --compatibility-date 2024-07-31 --experimental-assets assets"
+			);
+			expect(bodies.length).toBe(1);
+			expect(bodies[0]).toStrictEqual({
+				manifest: {
+					"/file-1.txt": {
+						hash: "0de3dd5df907418e9730fd2bd747bd5e",
+						size: 17,
+					},
+					"/boop/file-2.txt": {
+						hash: "7574a8cd3094a050388ac9663af1c1d6",
+						size: 17,
+					},
+				},
+			});
+		});
+
+		it("should upload an asset manifest of the files in the directory specified by [experimental_assets] config", async () => {
+			const assets = [
+				{ filePath: "file-1.txt", content: "Content of file-1" },
+				{ filePath: "boop/file-2.txt", content: "Content of file-2" },
+			];
+			writeAssets(assets);
+			writeWranglerToml({
+				experimental_assets: { directory: "assets" },
+			});
+			writeWorkerSource();
+			const bodies: AssetManifest[] = [];
+			msw.use(
+				http.post<never, AssetManifest>(
+					`*/accounts/some-account-id/workers/scripts/test-name/assets-upload-session`,
+					async ({ request }) => {
+						bodies.push(await request.json());
+						return HttpResponse.json(
+							{
+								success: true,
+								errors: [],
+								messages: [],
+								result: { jwt: "<<aus-completion-token>>", buckets: [] },
+							},
+							{ status: 201 }
+						);
+					}
+				)
+			);
+			// skips asset uploading since empty buckets returned
+			mockSubDomainRequest();
+			mockUploadWorkerRequest({
+				expectedExperimentalAssetsForm: [
+					["metadata", JSON.stringify({ assets: "<<aus-completion-token>>" })],
+				],
+			});
+			await runWrangler("deploy");
+			expect(bodies.length).toBe(1);
+			expect(bodies[0]).toStrictEqual({
+				manifest: {
+					"/file-1.txt": {
+						hash: "0de3dd5df907418e9730fd2bd747bd5e",
+						size: 17,
+					},
+					"/boop/file-2.txt": {
+						hash: "7574a8cd3094a050388ac9663af1c1d6",
+						size: 17,
+					},
+				},
+			});
+		});
+
+		it("should upload assets in the requested buckets", async () => {
+			const assets = [
+				{ filePath: "file-1.txt", content: "Content of file-1" },
+				{ filePath: "boop/file-2.txt", content: "Content of file-2" },
+				{ filePath: "boop/file-3.txt", content: "Content of file-3" },
+				{ filePath: "file-4.txt", content: "Content of file-4" },
+				{ filePath: "beep/file-5.txt", content: "Content of file-5" },
+				{ filePath: "beep/file-6.txt", content: "Content of file-1" },
```

Comment:
Realizing that we haven't yet defined behavior for when there are wonky characters in file names yet: spaces, question marks, etc.

My gut says we should essentially only support things which equal themselves when we `decodeURIComponent(new URL(`http://fakehost/${filePath}`).pathname))` and error for anything else. Can we add that check in to start with, and then revisit once we've got more of ASW2 in place?

Reply:
Please can we capture these rules in the SPEC.

Reply:
Good point, we should definitely validate this - I might leave this out of this PR for now and add a ticket to milestone 2 to follow up on this when we come to a decision.